ipq8064.c

```diff
@@ -0,0 +1,230 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/* Copyright (c) 2022, Linaro Limited */
+
+#include <sys/mman.h>
+#include <err.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "debugcc.h"
+
+#define GCC_PHYS		0x900000
+
+#define PDM_CLK_NS_REG		0x2cc0
+#define XO4_CLK_DIV_4		BIT(3) | BIT(4)
+#define XO4_CLK_BRANCH_ENA	BIT(7)
+#define CLK_ROOT_ENA		BIT(11)
+#define CXO_SRC_BRANCH_ENA	BIT(13)
+
+#define RINGOSC_NS_REG		0x2dc0
+#define RO_CLK_BRANCH_ENA	BIT(9)
+#define RO_ROOT_ENA		BIT(11)
+
+#define RINGOSC_TCXO_CTL_REG	0x2dc4
+#define RINGOSC_STATUS_REG	0x2dcc
+
+#define CLK_TEST_REG		0x2fa0
+#define RING_OSC_DBG_SEL	BIT(26) // Select cc_dbg_hs_clk instead of default value cc_ringosc_clk
+#define TEST_BUS_ENA		BIT(23)
+#define TEST_BUS_SEL_MASK	0x780000 // GENMASK(22, 19)
+#define HS_DBG_CLK_BRANCH_ENA	BIT(17)
+#define DBG_CLK_HS_SEL_MASK	0x1fc00 // GENMASK(16, 10)
+#define DBG_CLK_HS_SEL_SHIFT	10
+#define DBG_CLK_HS_SEL(x)	((x) << DBG_CLK_HS_SEL_SHIFT)
+#define LS_DBG_CLK_BRANCH_ENA	BIT(8)
+#define DBG_CLK_LS_SEL_MASK	0xff // GENMASK(7, 0)
+
+#define cc_dbg_ls_out_clk	DBG_CLK_HS_SEL(0x43)
+#define sc_dbg_hs1_clk		DBG_CLK_HS_SEL(0x41)
+
+#define TEST_PER_LS(x)		(HS_DBG_CLK_BRANCH_ENA | cc_dbg_ls_out_clk | LS_DBG_CLK_BRANCH_ENA | (x))
```

Comment:
It almost feels like the LS and HS branches should be turned into interim muxes.

Reply:
I wanted to prevent leaf madness but maybe it's not that bad. Will try

Reply:
Let's see if this works out or not.